File: algorithms/graph/counting-triangles/cpp/counting_triangles.cpp
#include "counting_triangles.h"
#include <vector>

int counting_triangles(const std::vector<int>& arr) {
    if (arr.size() < 2) return 0;
    int n = arr[0];
    int m = arr[1];
    
    if (arr.size() < 2 + 2 * m) return 0;
    if (n < 3) return 0;
    
    std::vector<std::vector<bool>> adj(n, std::vector<bool>(n, false));
    for (int i = 0; i < m; i++) {
        int u = arr[2 + 2 * i];
        int v = arr[2 + 2 * i + 1];
        if (u >= 0 && u < n && v >= 0 && v < n) {
            adj[u][v] = true;
            adj[v][u] = true;
        }
    }
    
    int count = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (adj[i][j]) {
                for (int k = j + 1; k < n; k++) {
                    if (adj[j][k] && adj[k][i]) {
                        count++;
                    }
                }
            }
        }
    }
    
    return count;
}

File: algorithms/graph/counting-triangles/go/counting_triangles.go
package countingtriangles

func CountingTriangles(arr []int) int {
	if len(arr) < 2 {
		return 0
	}
	n := arr[0]
	m := arr[1]

	if len(arr) < 2+2*m {
		return 0
	}
	if n < 3 {
		return 0
	}

	adj := make([][]bool, n)
	for i := range adj {
		adj[i] = make([]bool, n)
	}

	for i := 0; i < m; i++ {
		u := arr[2+2*i]
		v := arr[2+2*i+1]
		if u >= 0 && u < n && v >= 0 && v < n {
			adj[u][v] = true
			adj[v][u] = true
		}
	}

	count := 0
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if adj[i][j] {
				for k := j + 1; k < n; k++ {
					if adj[j][k] && adj[k][i] {
						count++
					}
				}
			}
		}
	}

	return count
}
